server/src/main/java/org/elasticsearch/action/search/SearchContextId.java

```diff
@@ -58,12 +59,33 @@ public boolean contains(ShardSearchContextId contextId) {
     public static BytesReference encode(
         List<SearchPhaseResult> searchPhaseResults,
         Map<String, AliasFilter> aliasFilter,
-        TransportVersion version
+        TransportVersion version,
+        ShardSearchFailure[] shardFailures
     ) {
         try (var out = new BytesStreamOutput()) {
             out.setTransportVersion(version);
             TransportVersion.writeVersion(version, out);
-            out.writeCollection(searchPhaseResults, SearchContextId::writeSearchPhaseResult);
+            boolean allowNullContextId = out.getTransportVersion().onOrAfter(TransportVersions.ALLOW_PARTIAL_SEARCH_RESULTS_IN_PIT);
+            int shardSize = searchPhaseResults.size() + (allowNullContextId ? shardFailures.length : 0);
+            out.writeVInt(shardSize);
+            for (var searchResult : searchPhaseResults) {
+                final SearchShardTarget target = searchResult.getSearchShardTarget();
+                target.getShardId().writeTo(out);
+                new SearchContextIdForNode(target.getClusterAlias(), target.getNodeId(), searchResult.getContextId()).writeTo(out);
+            }
+            if (allowNullContextId) {
+                /**
+                 * Shard failures are not encoded if there are nodes in the cluster that have not yet been upgraded to
+                 * {@link TransportVersions.ALLOW_PARTIAL_SEARCH_RESULTS_IN_PIT}.
+                 * These shards will be silently ignored during searches using this PIT; however, this situation should never occur,
+                 * as failures are not permitted when creating a point in time with versions prior to
+                 * {@link TransportVersions.ALLOW_PARTIAL_SEARCH_RESULTS_IN_PIT}.
```

Comment:
Agreed, I pushed https://github.com/elastic/elasticsearch/pull/111516/commits/b151ab17dcbbaafca68bdfb2b7dc71ed431712d0 to prevent the creation of the PIT with `allowPartialSearchResults=true` on a mixed cluster.